Weyland destinations lookup

wey-chiang opened the pull request:
## Background

Addresses two issues: #2640 and #1424 

The original generated alert fields implementation required users to use a destination's UUID. These changes allow users to use the destination display name in the `destinations` function.

There was also a logical flaw introduced in the rules engine `Rule:_run_command` function in which lists were not being properly validated -- the fix for that issue is included in these changes.

## Changes

- Added another conditional to the alert-delivery api for dynamic destinations to support display names.
- Refactored alert-forwarder API to use the rule OutputIDs for the alert OutputIDs and the alert dedup destinations for the alert destinations to be more aligned with the changes to the alert-deliver API
- Fixed logic in Rules Engine `Rule:_run_command` 


## Testing

- `mage test:ci`
- Manual deployment into `us-east-1`
  - Created rule that had `destinations` defined as follows
```
def destinations(event):
    return ["Weyland-Email-Test"]
```
  - Confirmed that the generated alert was properly routed to the destination 
  - Confirmed I did get an email for this destination (the dest. is an AWS SNS email subscription)

kostaspap commented on api/lambda/delivery/models/api.go:
```diff
@@ -162,7 +162,7 @@ type Alert struct {
 
 	// Destinations is the custom user-provided destination override
-	Destinations []string `json:"destinations,omitempty" validate:"dive,uuid4"`
+	Destinations []string `json:"destinations,omitempty"`
```
So the destinations can be either UUID or a Name? This seems very hacky API design. The API should be making clear what's included in it, if it's names or IDs, we cannot do guesswork on the content ^^ 

One way to address this is to have the rules engine due to name-> id resolution. When the rules engine loads the rules, it can also load the list of available outputs. It keeps a mapping of Names -> IDs. 
It can then resolve the Name to an ID and that's it. We don't have to change anything downstream. 

It also seems that we should change the public API of the rule override functions so that users know to use the names instead of UUIDs.

wey-chiang replied:
I was hesitant to include it in the rules engine due to potential performance concerns.

I like the idea however, and it would be more useful for testing logic i.e. checking if passed in destinations are actual existing destinations.

kostaspap replied:
> I was hesitant to include it in the rules engine due to potential performance concerns.
> 
> I like the idea however, and it would be more useful for testing logic i.e. checking if passed in destinations are actual existing destinations.

Note that there is already some I/O when we load the rules and the data models. I think  you could follow the same approach for outputs as well. 
See for example how we keep a cache of Rules: https://github.com/panther-labs/panther/blob/a4adb0e7498d89bf1cc6d1ae0138278e745f87d7/internal/log_analysis/rules_engine/src/engine.py#L43
and how we populate it when the `populate_rules()` is called: https://github.com/panther-labs/panther/blob/a4adb0e7498d89bf1cc6d1ae0138278e745f87d7/internal/log_analysis/rules_engine/src/engine.py#L160

I think you could follow the same approach for outputs. You could have a mapping of output name -> output id in memory and just pass this mapping to the rule or to the rule run method.

nhakmiller replied:
Yeah I think usually when we do "remappings" to make life easier for users, we try to do it as early as possible and not propagate that complexity through the pipeline. For example the `analysis-api` handles a lot of this work when new rules/policies are being created from `BulkUpload` so that the rest of the system doesn't have to think about it. I think a similar approach here would be good, having the rules-engine generate a lookup table one time and then referencing that as needed.

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -505,7 +505,7 @@ def _run_command(self, function: Callable, event: PantherEvent, expected_type: A
                 )
         else:
-            if not isinstance(expected_type, list) or not all([isinstance(x, (str, bool)) for x in result]):
+            if not isinstance(result, list) or not all([isinstance(x, (str, bool)) for x in result]):
```
Can you add a unit test that: a) Fails with old code b) Passes with new code, to make sure that the issue is fixed?

kostaspap replied:
Ah, nvm. I see that the `test_rule_with_invalid_destinations_type` is meant to do that :0

georgepsarakis commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -505,7 +505,7 @@ def _run_command(self, function: Callable, event: PantherEvent, expected_type: A
                 )
         else:
-            if not isinstance(expected_type, list) or not all([isinstance(x, (str, bool)) for x in result]):
+            if not isinstance(result, list) or not all([isinstance(x, (str, bool)) for x in result]):
```
It might help discern the different type checks if you use identity check for the expected type (Line 497):

```python
if expected_type is list:
    # check that result is instance of list and each element is either boolean or string
else:
    # check that result is instance of expected type
```

This would also actually allow passing multiple acceptable types in `expected_type`, which could be an iterable as well, since `isinstance` accepts that.

wey-chiang commented:
Changed the implementation to have the functionality restricted to the rules-engine and reverted downstream changes.

Since we've moved this into the rules-engine, we can now give better response to users as to invalid destinations, I've included some examples :) 

![image](https://user-images.githubusercontent.com/71197790/108400151-e5ff5780-71e8-11eb-8d76-b453701812e4.png)

![image](https://user-images.githubusercontent.com/71197790/108400167-ea2b7500-71e8-11eb-9de3-6c639773a6ed.png)

nhakmiller commented on internal/log_analysis/rules_engine/src/outputs_api.py:
```diff
@@ -0,0 +1,39 @@
+        get_input: Dict[str, Any] = {'getOutputs': {}}
+
+        response = self.client.invoke(FunctionName='panther-outputs-api', Payload=json.dumps(get_input).encode('utf-8'))
+        gateway_response = json.loads(response['Payload'].read())
```
Strictly speaking I believe this is a straight up lambda response, not an API gateway response. I would recommend re-naming this variable to `lambda_response`.

wey-chiang replied:
Makes sense, renamed it accordingly.

wey-chiang commented:
Some more SS from testing:

Note how there is no destination configured (destination set dynamically)

![image](https://user-images.githubusercontent.com/71197790/108410228-24027880-71f5-11eb-8786-60b8614280cf.png)

The Alerts page properly shows the destination now

![image](https://user-images.githubusercontent.com/71197790/108410251-2ebd0d80-71f5-11eb-88fa-b968efbc80ba.png)
![image](https://user-images.githubusercontent.com/71197790/108410265-32509480-71f5-11eb-87fe-a51a10eeefc5.png)

Email confirming the destinations are working as intended

![image](https://user-images.githubusercontent.com/71197790/108410348-52805380-71f5-11eb-8023-f486a48c9f4a.png)

nhakmiller commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -364,14 +365,38 @@ def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool
-        if len(destinations) > MAX_DESTINATIONS_SIZE:
+        invalid_destinations = []
+        # Map display names to destinations
+        outputs_display_names = {v.destination_display_name: v for k, v in outputs.items()} \
```
It seems this mapping/list could be calculated once when the destinations are initially loaded are re-used instead of being calculated for each alert.

wey-chiang replied:
It's ultimately the difference between adding two parameters versus adding one -- I decided on the side of simplicity at the cost of a few more cycles.

While we could just pack those into a list to only have one new parameter added, I figured this implementation would be more easily readable/understood

kostaspap replied:
I agree with Nick, we should do it only once. This code will run over billions of events per day so it is performance sensitive code. 
It can still be readable though, I don't think we are facing such a dilemma here. You could just pass the outputs e.g. in the rule constructor. Then the rule creates the following:
1. One map of name -> UUID
2. One set of UUIDs. 

When the rule runs the `destinations` function, it checks every item in the returned list. For each item, look up the map. If the lookup fails, check if the value exists in the set. If only few of the destinations defined by the user exist, we should just send notifications to those. If none of the destinations defined by the user exist, we just send to the default outputs.

wey-chiang replied:
I've moved the `outputs_display_names` into the engine initialization so it is only run once and initialized the same time as the destinations are populated.

nhakmiller reviewed (approved):
A bit concerned about the continually growing permissions of the rules-engine, but I don't think that's in scope for this PR.

kostaspap reviewed (commented):
This PR has the label `sync-to-release`. This doesn't seem right. The 1.16-RC is already out and has been tested. Adding this should require a new round of tests to make sure there is no regression in the way the rules work. My opinion is to remove the label. We can make a 1.16.1 if someone needs that urgently.

wey-chiang commented:
Huh, CircleCI is failing because

> # Run  npm install aws-amplify@3.3.20  to resolve 1 vulnerability

Will include that package update in this PR

wey-chiang commented:
After updating `aws-amplify` it looks like `npm-audit` is still showing 9 high vulns, would updating these libraries to a newer version break anything?@nhakmiller 

If not, I'll include those in the PR as well

nhakmiller commented:
> After updating `aws-amplify` it looks like `npm-audit` is still showing 9 high vulns, would updating these libraries to a newer version break anything?@nhakmiller
> 
> If not, I'll include those in the PR as well

I'd say go for it. But now the bundle size check is failing, not sure what's causing that.

wey-chiang commented:
Ran into an issue testing which revealed an issue that was not caught by unit tests. 
I've added some more unit tests to better test the feature.

wey-chiang commented:
Removed the npm package upgrades

@kostaspap those will be blocking future CI tests and need to be addressed -- I tried to upgrade them but upgrading all of the packages presenting high vulns breaks deployment

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -364,14 +370,36 @@ def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool
+        if invalid_destinations:
+            if use_default_on_exception:
+                self.logger.warning('destinations method yielded invalid destinations: %s', str(invalid_destinations))
+                return []
```
Shouldn't we return None here? From my understanding, if we return `[]` then we will just not send notification to any destination.

wey-chiang replied:
Huh... yeah 👍

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -364,14 +370,36 @@ def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool
+
+        if invalid_destinations:
+            if use_default_on_exception:
+                self.logger.warning('destinations method yielded invalid destinations: %s', str(invalid_destinations))
```
I think this should be just `info` level. We don't want to get paged for this

wey-chiang replied:
I'll make those changes for the rest of the generated alert fields as well then.

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -364,14 +370,36 @@ def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool
+
+        if len(standardized_destinations) > MAX_DESTINATIONS_SIZE:
             # If generated field exceeds max size, truncate it
             self.logger.warning(
```
Same here, i think this should be info level

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -364,14 +370,36 @@ def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool
             )
-            return destinations[:MAX_DESTINATIONS_SIZE]
-        return destinations
+            return standardized_destinations[:MAX_DESTINATIONS_SIZE]
```
Have we documented these limits?

wey-chiang replied:
No, I'll queue that up to mention that in the `Alert Field Overrides` doc

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -364,14 +370,36 @@ def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool
-        if len(destinations) > MAX_DESTINATIONS_SIZE:
+        # Check for (in)valid destinations
+        invalid_destinations = []
+        standardized_destinations = []
```
We want them to be unique, right? This could be a set. it could simplify the logic below as well.

wey-chiang replied:
Sure, yeah they could be a set but I don't see how that would simplify the logic below.

kostaspap replied:
It can be rewritten like the following:
```
for dest in destinations:
    if dest in outputs_display_names: # this means the value returned by the destination function was a destination name
        standardized_destinations.add(outputs_display_names[dest].destination_id)   
     elif dest in outputs: # this means the value returned by the destination function was a destination id. 
        standardized_destinations.add(dest)
     else:
         invalid_destinations.add(dest)
```
Sets are ideal for deduplication like the one you are doing here.

kostaspap commented:
> Removed the npm package upgrades
> 
> @kostaspap those will be blocking future CI tests and need to be addressed -- I tried to upgrade them but upgrading all of the packages presenting high vulns breaks deployment

The FE team is responsible for updating those ones - @3nvi  I think is already aware of it.

kostaspap commented on internal/log_analysis/rules_engine/src/engine.py:
```diff
@@ -24,26 +24,32 @@
         self._last_update = datetime.utcfromtimestamp(0)
         self.log_type_to_data_models: Dict[str, DataModel] = collections.defaultdict()
         self.log_type_to_rules: Dict[str, List[Rule]] = collections.defaultdict(list)
+        self.destinations: Dict[str, Destination] = collections.defaultdict()
```
Does this contain a mapping of id -> destination? Could you rename it to `id_to_destination` so that the name follows the same pattern as below?

kostaspap commented on internal/log_analysis/rules_engine/src/rule.py:
```diff
@@ -351,7 +353,11 @@ def _get_description(self, event: PantherEvent, use_default_on_exception: bool =
-    def _get_destinations(self, event: PantherEvent, use_default_on_exception: bool = True) -> Optional[List[str]]:
+    def _get_destinations(self,
+                          event: PantherEvent,
+                          outputs: dict,
```
nit: The `outputs` are `output_ids` right? Might be good to rename, makes it easier for the reader.
